packages/block-editor/src/components/global-styles/use-global-styles-output.js

```diff
@@ -395,17 +395,40 @@ export function getStylesDeclarations(
 	);
 
 	/*
-	 * Set background defaults.
-	 * Applies to all background styles except the top-level site background.
+	 * Preprocess background image values.
+	 * 1. Sets default values for blocks (not root).
+	 * 2. .
+	 *
+	 * Note: A refactor is for the style engine to handle ref resolution (and possibly defaults)
+	 * via a public util used internally and externally. Theme.json tree and defaults could be passed
+	 * as options.
 	 */
-	if ( ! isRoot && !! blockStyles.background ) {
-		blockStyles = {
-			...blockStyles,
-			background: {
-				...blockStyles.background,
-				...setBackgroundStyleDefaults( blockStyles.background ),
-			},
-		};
+	if ( !! blockStyles.background ) {
+		// 1. Set default values for block styles except the top-level site background
+		if ( ! isRoot ) {
+			blockStyles = {
+				...blockStyles,
+				background: {
+					...blockStyles.background,
+					...setBackgroundStyleDefaults( blockStyles.background ),
+				},
+			};
+		}
+
+		/*
+		 * 2. Resolve dynamic values before they are compiled by the style engine,
+		 * which doesn't (yet) resolve dynamic values.
+		 */
+		if ( blockStyles.background?.backgroundImage?.ref ) {
```

Comment:
A great deal of custom code just for background images.

It works, but could be reduce by moving ref resolution into the style engine. The condition is that we'll need to pass a theme.json object to the style engine methods.